Hardware/Nic/NicGroup.cs

```diff
@@ -20,10 +22,39 @@ public NicGroup(ISettings settings)
 
         private void ScanNics(ISettings settings)
         {
-            NetworkInterfaces = NetworkInterface.GetAllNetworkInterfaces();
-            _hardware = NetworkInterfaces.Where(DesiredNetworkType)
-                .Select((x, i) => new Nic(x, settings, i))
-                .ToList();
+            // If no network is marked up (excluding loopback and tunnel) then don't scan
+            // for interfaces.
+            if (!NetworkInterface.GetIsNetworkAvailable())
+            {
+                return;
+            }
+
+            // When multiple events fire concurrently, we don't want threads interferring
+            // with others as they manipulate non-thread safe state.
+            lock (_scanLock)
+            {
+                NetworkInterfaces = NetworkInterface.GetAllNetworkInterfaces()
+                    .Where(DesiredNetworkType)
+                    .OrderBy(x => x.Name)
+                    .ToArray();
+
+                var scanned = NetworkInterfaces.ToDictionary(x => x.Id, x => x);
+                var newNics = scanned.Where(x => !_nics.ContainsKey(x.Key));
+                var removedNics = _nics.Where(x => !  scanned.ContainsKey(x.Key)).ToList();

```

Comment:
whitespace

Reply:
also doesn't need to use ToList

Reply:
I believe creating the temporary list to remove items in a collection is necessary, as removing an item in the midst of an iterator will invalidate the iterator and throw an exception: https://stackoverflow.com/q/3506121/433785